src/blocks/PixelArtBlock.tsx

```diff
@@ -0,0 +1,269 @@
+      <div style={{ display: 'flex', flexDirection: 'column', maxWidth: isEditMode ? '500px' : '2000px', width: '100%'}}>
+        {isEditMode &&
+          <Stack direction='row' spacing={2} paddingBottom={1} justifyContent='center'>
+            <div>
+              <label>Set Color: </label>
+              <input
+                type='color'
+                id='colorInput'
+                value={color}
+                onChange={(e) => setColor(e.target.value)}
+              />
+            </div>
+            <div>
+              <label>Show Guide: </label>
+              <input
+                type='checkbox'
+                id='toggleGrid'
+                checked={showGrid}
+                onChange={() => {setShowGrid(!showGrid)}}
+              />
+            </div>
+            <div>
+              <button type='button' id='clearButton' onClick={clearCanvas}>
+                Clear
+              </button>
+            </div>
+          </Stack>
+        }
+        <canvas
+          ref={canvasRef}
+          width={width}
+          height={height}
+          style={{ cursor: 'pointer '}}
+          onMouseDown={handleCanvasClick}
+        />
+        {isEditMode &&
+          <Button
+            type='submit'
+            variant='contained'
+            className='save-modal-button'
+            sx={{ mt: 3, mb: 2 }}
+            onClick={savePixelState}
+          >
+            Save
+          </Button>
+        }
+      </div>
+    </div>
+  )
+}
+
+export default class PixelArtBlock extends Block {
+  render() {
+    if (Object.keys(this.model.data).length === 0) {
+      return BlockFactory.renderEmptyState(this.model, this.onEditCallback!)
+    }
+
+    const {
+      numPixelsPerSide,
+      pixelsArrStringified,
+    } = this.model.data;
+    console.log('loading pixels', pixelsArrStringified);
+    const pixels = JSON.parse(pixelsArrStringified);
+
+    return (
+      <PixelCanvas
+        numPixelsPerSide={parseInt(numPixelsPerSide)}
+        isEditMode={false}
+        initialPixels={pixels}
+      />
+    );
+  }
+
+  renderEditModal(done: (data: BlockModel) => void) {
+    const numPixels = 5;
+    const {
+      numPixelsPerSide,
+      pixelsArrStringified,
+    } = this.model.data;
+    console.log('loading pixels', pixelsArrStringified);
+
+    const onSave = (pixels: string[][]) => {
+      this.model.data['numPixelsPerSide'] = numPixels.toString();
+      this.model.data['pixelsArrStringified'] = JSON.stringify(pixels);
+      console.log('saving pixels: ', this.model.data['pixelsArrStringified']);
+      done(this.model);
+    }
+    return (
+      <PixelCanvas
+        numPixelsPerSide={(numPixelsPerSide && parseInt(numPixelsPerSide)) || numPixels}
+        initialPixels={pixelsArrStringified && JSON.parse(pixelsArrStringified)}
+        isEditMode={true}
+        onSave={onSave}
+      />
+    )
+  }
+
+  renderErrorState() {
```

Comment:
Does this get called automatically within the block framework, or must it be called explicitly? @nconfrey

Reply:
It must be called explicitly for errors that you know your block throws. Additionally, the block framework will call it if the network data is corrupted somehow, but this is a rare occurrence.